### PR TITLE
chore(serialization): Improve deserialization errors

### PR DESCRIPTION
**Pull request by adambollen**
### Description

BT-5115

Pursuant to the ticket, we want to make deserialization errors easier to reason about; the current changes are based on a similar fix in the `fauna-jvm` driver here: https://github.com/fauna/fauna-jvm/pull/129

The change ultimately is an improvement on the `SerializationException` message to avoid spitting out the unexpected token received on the wire from Fauna and instead mapping that token to the expected `FaunaType` (a new enum) that should help point consumers to the issue in their logic where they are expecting to deserialize as one type but the .NET driver's serialization logic saw a different type.  Tests have been updated and a couple new ones were added.

### Motivation and context

See above.

### How was the change tested?

I used new test automation to repeatedly validate the changes with various tweaks until I liked the final product.  I also updated the existing serialization automation with the new exception messaging and everything is green.

(I also made heavy use of the new debug logging that I just merged. 😄 )

### Change types

* - [x] Bug fix (non-breaking change that fixes an issue)
* - [x] New feature (non-breaking change that adds functionality)
* - [ ] Breaking change (backwards-incompatible fix or feature)

### Checklist:

* - [x] My code follows the code style of this project.
* - [x] My change requires a change to Fauna documentation.
* - [ ] My change requires a change to the README, and I have updated it accordingly.

----
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by adambollen** [comment]
@jrodewig , not sure if this requires any updates to docs - only if we show the `SerializationException` message output when deserialization fails.  FYI.

File: Fauna/Serialization/TokenType.cs [hunk 1]
```diff
@@ -66,3 +65,59 @@ public enum TokenType
...
+            case TokenType.EndRef:
+                return FaunaType.Ref;
+
+            case TokenType.StartDocument:
+            case TokenType.EndDocument:
+                return FaunaType.Document;
+
+            case TokenType.String:
+                return FaunaType.String;
+            // BUGBUG: @bytes support is missing; BT-5183
```
Comment (cynicaljoy) [inline Fauna/Serialization/TokenType.cs:97]:
```suggestion
            // BUG: @bytes support is missing; BT-5183
```

nit

File: Fauna/Serialization/Serializer.cs [hunk 2]
```diff
@@ -102,9 +102,18 @@ public static ISerializer Generate(MappingContext context, Type targetType)
...
         {
+            var args = targetType.GetGenericArguments();
+            var blocked = args.Where(arg => arg.GetInterfaces().Contains(typeof(IQueryFragment)));
+
+            if (blocked.Any())
+            {
+                // throw new ArgumentException(
+                //     $"Query types ({string.Join(", ", blocked.Select(x => x.FullName))}) are not supported as " +
+                //     $"generic type arguments; use {nameof(QueryArr)} or {nameof(QueryObj)} when composing complex queries.");
+            }
```
Comment (cynicaljoy) [inline Fauna/Serialization/Serializer.cs:113]: TODO?

File: Fauna/Serialization/ListSerializer.cs [hunk 3]
```diff
@@ -51,6 +53,13 @@ public override void Serialize(MappingContext ctx, Utf8FaunaWriter w, object? o)
             o.GetType().GetGenericTypeDefinition() == typeof(List<>) ||
             o.GetType().GetGenericTypeDefinition() == typeof(IEnumerable))
         {
+            var genType = o.GetType().GenericTypeArguments.SingleOrDefault();
+
+            if (genType is not null && genType.GetInterfaces().Contains(typeof(IQueryFragment)))
+            {
+                //throw new ArgumentException($"{genType} cannot be serialized in a List<>; try {nameof(QueryArr)} instead.");
```
Comment (cynicaljoy) [inline Fauna/Serialization/ListSerializer.cs:60]: TODO?

File: Fauna/Serialization/DictionarySerializer.cs [hunk 4]
```diff
@@ -36,6 +38,14 @@ public override void Serialize(MappingContext context, Utf8FaunaWriter writer, o
                 writer.WriteNullValue();
                 break;
             case Dictionary<string, T> d:
+                // Get generic type arg for the value, which should be the 2nd type arg
+                var genType = d.GetType().GenericTypeArguments[1];
+
+                if (genType is not null && genType.GetInterfaces().Contains(typeof(IQueryFragment)))
+                {
+                    //throw new ArgumentException($"{genType} cannot be serialized in a Dictionary<>; try {nameof(QueryObj)} instead.");
```
Comment (cynicaljoy) [inline Fauna/Serialization/DictionarySerializer.cs:46]: TODO?

Comment (adambollen) [reply Fauna/Serialization/DictionarySerializer.cs:46]: Haha, yep, that was my failed attempt at the other fix.  I'll clean that out.

**Review by cynicaljoy — APPROVED** [review]

**Pull request merged**